{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "83d92116-fc6a-437e-a6e4-61517c2fb1f3",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Correlated Phenotypes Simulation\n",
    "\n",
    "## Goal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7485dd87-86ab-4079-90f1-108e4ac9d7b8",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Here we used estimated correlation to simulate phenotype (traits) with residual correlation (see details in Supplementary Note S.6.2 of ColocBoost paper).\n",
    "\n",
    "$\\mathbf{Y}_1 = \\mathbf{X} \\boldsymbol{\\beta}_1 + \\boldsymbol{\\varepsilon} \\quad \\text{with} \\ \\boldsymbol{\\varepsilon} \\sim \\mathcal{N}(\\mathbf{0}, \\sigma_y^2 \\boldsymbol{\\Sigma})$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b9ba7a4-91e5-4a5d-9185-b3219a507e11",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "Input and output format are the same as other phenotype simulation files. $\\Sigma$ is estimated from MASH using AD-xQTL traits."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74e49930-ad7c-46b2-bfda-7bf97e84fdb3",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Phenotype simulation code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93d23341-1fa1-402d-9bb6-fd63897adbdc",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "outputs": [],
   "source": [
    "[real_simulation_2trait_corr]\n",
    "parameter: genofile = paths\n",
    "# pheno_file: give genotype file (in plink)，we can read the gentype matrix. These files are separated by TADs.\n",
    "parameter: cwd = path(\"output\")\n",
    "parameter: job_size = 30\n",
    "parameter: walltime = \"100h\"\n",
    "parameter: mem = \"30G\"\n",
    "parameter: numThreads = 1\n",
    "# specify the number of causal variants\n",
    "parameter: n_trait = 2\n",
    "parameter: n_causal = 1\n",
    "parameter: h2g = 0.05\n",
    "parameter: total_h2g = False\n",
    "parameter: share_pattern = \"all\"\n",
    "parameter: independent = False\n",
    "# specify the number of traits (phenotypes)\n",
    "parameter: container = \"\"\n",
    "\n",
    "input: genofile, group_by = 1\n",
    "output: f'{cwd:a}/{step_name}/sample_{_index}_real_simulation_{n_causal}_ncausal_{n_trait}_trait.rds'\n",
    "\n",
    "R: expand = '${ }', stdout = f\"{_output:n}.stdout\", stderr = f\"{_output:n}.stderr\", container = container\n",
    "    # Here we use estimated correlation to simulate traits with residual correlation\n",
    "    # --- Load necessary libraries ---\n",
    "    library(\"MASS\")\n",
    "    library(\"plink2R\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"tidyverse\")\n",
    "  \n",
    "    # install simulation package\n",
    "    # devtools::install_github(\"StatFunGen/simxQTL\", build_vignettes = FALSE)\n",
    "    # BiocManager::install(\"StatFunGen/pecotmr\")\n",
    "    library(\"pecotmr\")\n",
    "    library(\"simxQTL\")\n",
    "\n",
    "    # --- Read genotype ---\n",
    "    simu_file = ${_input:r}\n",
    "    geno <- read_plink(${_input:nr})\n",
    "    gene_name = str_extract(simu_file, \"ENSG[0-9]+\")\n",
    "    gene_tss_map = read_tsv(\"./Data/gene_cis_TADB_mapper.tsv\")\n",
    "\n",
    "    # --- Filter by TSS window (+/- 1.5Mb) ---\n",
    "    TSS_pos = gene_tss_map$TSS[which(gene_tss_map$gene_id == gene_name)][1]\n",
    "    keep_index = which(geno$bim$V4 > (TSS_pos - 1500000) & geno$bim$V4 < (TSS_pos + 1500000))\n",
    "    geno$bed = geno$bed[, keep_index]\n",
    "\n",
    "    # --- SNP filtering ---\n",
    "    imiss = 0.1\n",
    "    maf = 0.05\n",
    "    Xmat = filter_X(geno$bed, imiss, maf)\n",
    "\n",
    "    # --- Select causal variants ---\n",
    "    ncausal = ${n_causal}\n",
    "    ntrait = ${n_trait}\n",
    "    indep = ${\"TRUE\" if independent else \"FALSE\"}\n",
    "\n",
    "    if (indep) {\n",
    "        LD_vars = 1  # Initialize LD check\n",
    "\n",
    "        if (ncausal == 1) {\n",
    "            # Only one causal variant needed\n",
    "            vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "        } else {\n",
    "            # Ensure selected variants are approximately independent (LD < 0.3)\n",
    "            while (length(LD_vars != 0)) {\n",
    "                vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "                cor_mat = cor(Xmat[, vars])\n",
    "                LD_vars = which(colSums(abs(cor_mat) > 0.3) > 1)\n",
    "            }\n",
    "        }\n",
    "    } else {\n",
    "        LD_vars = 1  # Initialize LD check\n",
    "\n",
    "        if (ncausal == 1) {\n",
    "            vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "        } else {\n",
    "            # Avoid perfectly correlated variants (|cor| = 1)\n",
    "            while (length(LD_vars != 0)) {\n",
    "                vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "                cor_mat = cor(Xmat[, vars])\n",
    "                LD_vars = which(colSums(abs(cor_mat) == 1) > 1)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    # --- Set up effect sizes ---\n",
    "    B = matrix(0, nrow = ncol(Xmat), ncol = ntrait)\n",
    "\n",
    "    for (i in 1:2) {\n",
    "        causal_index = vars\n",
    "        beta = sim_beta_fix_variant(G = Xmat, causal_index = causal_index, is_h2g_total = FALSE)\n",
    "        B[, i] = beta\n",
    "    }\n",
    "\n",
    "    # --- Record causal variants ---\n",
    "    variant = list()\n",
    "    for (i in 1:ncol(B)) {\n",
    "        variant[[i]] = which(B[, i] != 0)\n",
    "    }\n",
    "\n",
    "    # --- Simulate phenotype with estimated residual covariance ---\n",
    "    variance = readRDS(\"./Data/vhat_ROSMAP_eQTL.rds\")\n",
    "    two_trait_var = variance[c(\"AC_DeJager_eQTL\", \"PCC_DeJager_eQTL\"), c(\"AC_DeJager_eQTL\", \"PCC_DeJager_eQTL\")]\n",
    "\n",
    "    phenotype = sim_multi_traits(\n",
    "        G = Xmat,\n",
    "        B = B,\n",
    "        h2g = ${h2g},\n",
    "        is_h2g_total = ${\"TRUE\" if total_h2g else \"FALSE\"},\n",
    "        residual_corr = two_trait_var\n",
    "    )\n",
    "    phenotype = phenotype$P\n",
    "\n",
    "    # --- Save results ---\n",
    "    data = list()\n",
    "    data[[\"X\"]] = Xmat\n",
    "    data[[\"Y\"]] = phenotype\n",
    "    data[[\"variant\"]] = variant\n",
    "    saveRDS(data, ${_output:r})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b68319-2c49-49b2-9ad1-d5a88f8f2bf7",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[real_simulation_5trait_corr]\n",
    "parameter: genofile = paths\n",
    "# pheno_file: give genotype file (in plink)，we can read the gentype matrix. These files are separated by TADs.\n",
    "parameter: cwd = path(\"output\")\n",
    "parameter: job_size = 30\n",
    "parameter: walltime = \"100h\"\n",
    "parameter: mem = \"30G\"\n",
    "parameter: numThreads = 1\n",
    "# specify the number of causal variants\n",
    "parameter: n_trait = 5\n",
    "parameter: n_causal = 1\n",
    "parameter: h2g = 0.05\n",
    "parameter: total_h2g = False\n",
    "parameter: independent = False\n",
    "parameter: share_pattern = \"all\"\n",
    "# specify the number of traits (phenotypes)\n",
    "parameter: container = \"\"\n",
    "input: genofile, group_by = 1\n",
    "output: f'{cwd:a}/{step_name}/sample_{_index}_real_simulation_{n_causal}_ncausal_{n_trait}_trait.rds'\n",
    "R:  expand = '${ }', stdout = f\"{_output:n}.stdout\", stderr = f\"{_output:n}.stderr\", container = container \n",
    "    # --- Load necessary libraries ---\n",
    "    library(\"MASS\")\n",
    "    library(\"plink2R\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"tidyverse\")\n",
    "  \n",
    "    # install simulation package\n",
    "    # devtools::install_github(\"StatFunGen/simxQTL\", build_vignettes = FALSE)\n",
    "    # BiocManager::install(\"StatFunGen/pecotmr\")\n",
    "    library(\"pecotmr\")\n",
    "    library(\"simxQTL\")\n",
    "  \n",
    "    # read the plink file\n",
    "    simu_file = ${_input:r}\n",
    "    geno <- read_plink(${_input:nr})\n",
    "    gene_name = str_extract(simu_file, \"ENSG[0-9]+\")\n",
    "    gene_tss_map = read_tsv(\"./Data/gene_cis_TADB_mapper.tsv\")\n",
    "    # filter by distance with. TSS\n",
    "    TSS_pos = gene_tss_map$TSS[which(gene_tss_map$gene_id == gene_name)][1]\n",
    "    keep_index = which(geno$bim$V4 > TSS_pos - 1500000 | geno$bim$V4 < TSS_pos + 1500000)\n",
    "    geno$bed = geno$bed[,keep_index]\n",
    "    # filter out columns with missing rate > 0.1\n",
    "    imiss = 0.1\n",
    "    # filter out columns with MAF < 0.05\n",
    "    maf = 0.05\n",
    "    Xmat = filter_X(geno$bed, imiss, maf)\n",
    "    # only keep the first 4000 variants\n",
    "    ncausal = ${n_causal}\n",
    "    ntrait = ${n_trait}\n",
    "    indep = ${\"TRUE\" if independent else \"FALSE\"}\n",
    "    if (indep) {\n",
    "        LD_vars = 1  # Initialize LD check\n",
    "\n",
    "        if (ncausal == 1) {\n",
    "            # Only one causal variant needed\n",
    "            vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "        } else {\n",
    "            # Ensure selected variants are approximately independent (LD < 0.3)\n",
    "            while (length(LD_vars != 0)) {\n",
    "                vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "                cor_mat = cor(Xmat[, vars])\n",
    "                LD_vars = which(colSums(abs(cor_mat) > 0.3) > 1)\n",
    "            }\n",
    "        }\n",
    "    } else {\n",
    "        LD_vars = 1  # Initialize LD check\n",
    "\n",
    "        if (ncausal == 1) {\n",
    "            vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "        } else {\n",
    "            # Avoid perfectly correlated variants (|cor| = 1)\n",
    "            while (length(LD_vars != 0)) {\n",
    "                vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "                cor_mat = cor(Xmat[, vars])\n",
    "                LD_vars = which(colSums(abs(cor_mat) == 1) > 1)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "  \n",
    "    prop = readRDS(\"/home/hs3393/cloud_colocalization/simulation_code/trait6_prop.rds\")[,1:4]\n",
    "    proportions = prop[ncausal,]\n",
    "    sampled_name <- as.numeric(sample(names(proportions), size =  ncausal, prob = proportions, replace = TRUE))\n",
    "    phenotype = list()\n",
    "    B = matrix(0, nrow =  ncol(Xmat), ncol = ntrait)\n",
    "    config = matrix(0, nrow =  ntrait, ncol = ncausal)\n",
    "    for(i in c(1:ncausal)){\n",
    "          coloc_trait = sample(c(1:ntrait), sampled_name[i])\n",
    "          config[coloc_trait, i] = 1\n",
    "    }\n",
    "    for(i in 1:nrow(config)){\n",
    "        beta = B[,i, drop = FALSE]\n",
    "        index = which(config[i,] == 1)\n",
    "        if(length(index) > 0){\n",
    "            causal_index = vars[index]\n",
    "            beta = sim_beta_fix_variant(G = Xmat, causal_index = causal_index, is_h2g_total = FALSE)\n",
    "            B[, i] = beta\n",
    "        }else{\n",
    "            B[, i] = beta\n",
    "        }\n",
    "    }\n",
    "  \n",
    "    variance = readRDS(\"./Data/vhat_ROSMAP_eQTL.rds\")\n",
    "    two_trait_var = variance[c(\"AC_DeJager_eQTL\", \"PCC_DeJager_eQTL\"), c(\"AC_DeJager_eQTL\", \"PCC_DeJager_eQTL\")]\n",
    "    five_trait_var <- variance[c(1:5), c(1:5)]\n",
    "    phenotype = sim_multi_traits(G = Xmat, B = B, h2g = ${h2g}, is_h2g_total = ${\"TRUE\" if total_h2g else \"FALSE\"}, residual_corr = five_trait_var)\n",
    "  \n",
    "    variant = list()\n",
    "    for(i in 1:ncol(B)){\n",
    "        variant[[i]] = which(B[,i] != 0)\n",
    "    }\n",
    "    X = Xmat\n",
    "    Y = phenotype$P\n",
    "    colnames(Y) = paste0(\"Trait\", c(1:ntrait))\n",
    "    data = list()\n",
    "    data[[\"X\"]] = Xmat\n",
    "    data[[\"Y\"]] = Y\n",
    "    data[[\"variant\"]] = variant\n",
    "    saveRDS(data, ${_output:r})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63b40a1-41c1-497f-98ee-ee11e7303f76",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[real_simulation_10trait_corr]\n",
    "parameter: genofile = paths\n",
    "# pheno_file: give genotype file (in plink)，we can read the gentype matrix. These files are separated by TADs.\n",
    "parameter: cwd = path(\"output\")\n",
    "parameter: job_size = 30\n",
    "parameter: walltime = \"100h\"\n",
    "parameter: mem = \"30G\"\n",
    "parameter: numThreads = 1\n",
    "# specify the number of causal variants\n",
    "parameter: n_trait = 10\n",
    "parameter: n_causal = 1\n",
    "parameter: h2g = 0.02\n",
    "parameter: total_h2g = False\n",
    "parameter: independent = False\n",
    "parameter: share_pattern = \"all\"\n",
    "# specify the number of traits (phenotypes)\n",
    "parameter: container = \"\"\n",
    "input: genofile, group_by = 1\n",
    "output: f'{cwd:a}/{step_name}/sample_{_index}_real_simulation_{n_causal}_ncausal_{n_trait}_trait.rds'\n",
    "R:  expand = '${ }', stdout = f\"{_output:n}.stdout\", stderr = f\"{_output:n}.stderr\", container = container \n",
    "    # --- Load necessary libraries ---\n",
    "    library(\"MASS\")\n",
    "    library(\"plink2R\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"tidyverse\")\n",
    "  \n",
    "    # install simulation package\n",
    "    # devtools::install_github(\"StatFunGen/simxQTL\", build_vignettes = FALSE)\n",
    "    # BiocManager::install(\"StatFunGen/pecotmr\")\n",
    "    library(\"pecotmr\")\n",
    "    library(\"simxQTL\")\n",
    "  \n",
    "    # read the plink file\n",
    "    simu_file = ${_input:r}\n",
    "    geno <- read_plink(${_input:nr})\n",
    "    gene_name = str_extract(simu_file, \"ENSG[0-9]+\")\n",
    "    gene_tss_map = read_tsv(\"./Data/gene_cis_TADB_mapper.tsv\")\n",
    "    # filter by distance with. TSS\n",
    "    TSS_pos = gene_tss_map$TSS[which(gene_tss_map$gene_id == gene_name)][1]\n",
    "    keep_index = which(geno$bim$V4 > TSS_pos - 1500000 | geno$bim$V4 < TSS_pos + 1500000)\n",
    "    geno$bed = geno$bed[,keep_index]\n",
    "    # filter out columns with missing rate > 0.1\n",
    "    imiss = 0.1\n",
    "    # filter out columns with MAF < 0.05\n",
    "    maf = 0.05\n",
    "    Xmat = filter_X(geno$bed, imiss, maf)\n",
    "    \n",
    "    ncausal = ${n_causal}\n",
    "    ntrait = ${n_trait}\n",
    "    indep = ${\"TRUE\" if independent else \"FALSE\"}\n",
    "    if (indep) {\n",
    "        LD_vars = 1  # Initialize LD check\n",
    "\n",
    "        if (ncausal == 1) {\n",
    "            # Only one causal variant needed\n",
    "            vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "        } else {\n",
    "            # Ensure selected variants are approximately independent (LD < 0.3)\n",
    "            while (length(LD_vars != 0)) {\n",
    "                vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "                cor_mat = cor(Xmat[, vars])\n",
    "                LD_vars = which(colSums(abs(cor_mat) > 0.3) > 1)\n",
    "            }\n",
    "        }\n",
    "    } else {\n",
    "        LD_vars = 1  # Initialize LD check\n",
    "\n",
    "        if (ncausal == 1) {\n",
    "            vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "        } else {\n",
    "            # Avoid perfectly correlated variants (|cor| = 1)\n",
    "            while (length(LD_vars != 0)) {\n",
    "                vars = sample(1:ncol(Xmat), size = ncausal)\n",
    "                cor_mat = cor(Xmat[, vars])\n",
    "                LD_vars = which(colSums(abs(cor_mat) == 1) > 1)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    B = matrix(0, nrow =  ncol(Xmat), ncol = ntrait)\n",
    "  \n",
    "    for(i in 1:2){\n",
    "        causal_index = vars\n",
    "        beta = sim_beta_fix_variant(G = Xmat, causal_index = causal_index, is_h2g_total = FALSE)\n",
    "        B[, i] = beta\n",
    "    }\n",
    "  \n",
    "    variant = list()\n",
    "    for(i in 1:ncol(B)){\n",
    "      variant[[i]] = which(B[,i] != 0)\n",
    "    }\n",
    "    variance = readRDS(\"./Data/vhat_ROSMAP_eQTL.rds\")\n",
    "    two_trait_var = variance[c(\"AC_DeJager_eQTL\", \"PCC_DeJager_eQTL\"), c(\"AC_DeJager_eQTL\", \"PCC_DeJager_eQTL\")]\n",
    "    five_trait_var <- variance[c(1:5), c(1:5)]\n",
    "    phenotype = sim_multi_traits(G = Xmat, B = B, h2g = ${h2g}, is_h2g_total = ${\"TRUE\" if total_h2g else \"FALSE\"}, residual_corr = variance)\n",
    "    phenotype = phenotype$P\n",
    "    X = Xmat\n",
    "    Y = phenotype\n",
    "    data = list()\n",
    "    data[[\"X\"]] = Xmat\n",
    "    data[[\"Y\"]] = phenotype\n",
    "    data[[\"variant\"]] = variant\n",
    "    saveRDS(data, ${_output:r})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce78ba89-a8cb-4b0f-bfdc-46e69001aaa0",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Bash submission"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "027edb4c-46e3-48aa-864f-852e396e74d2",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "work_dir=\"/home/hs3393/cb_Mar/simulation_data/simulation_correlated/\"\n",
    "job=\"real_simulation_2trait_corr\"\n",
    "mkdir -p ${work_dir}\n",
    "mkdir -p ${work_dir}/code\n",
    "mkdir -p ${work_dir}/log\n",
    "cd ${work_dir}/code\n",
    "\n",
    "# Create the base_script file and write the bash code into it\n",
    "cat << 'EOF' > base_script\n",
    "#!/bin/bash -l\n",
    "# NOTE the -l flag!\n",
    "#\n",
    "#SBATCH -t 30:00:00\n",
    "#SBATCH --mem=30000\n",
    "#SBATCH -J JOB\n",
    "#SBATCH -o WORK_DIR/log/JOB.\"%j\".out\n",
    "#SBATCH -e WORK_DIR/log/JOB.\"%j\".err\n",
    "\n",
    "source ~/mamba_activate.sh\n",
    "module load Singularity\n",
    "\n",
    "\n",
    "sos run /home/hs3393/cb_Mar/simulation_code/7_Correlated_Simulation.ipynb JOB \\\n",
    "    --genofile `ls /home/hs3393/cloud_colocalization/simulation_data/selected_genes_genotype/*.bim` \\\n",
    "    --n_causal CAUSAL --mem 30G --h2g 0.05 --independent \\\n",
    "    --cwd WORK_DIR/\n",
    "EOF\n",
    "\n",
    "base_script=\"base_script\"\n",
    "for causal in 1 2 3; do\n",
    "    output_script=\"causal_${causal}.sh\"\n",
    "    cat ${base_script}|sed \"s|CAUSAL|${causal}|g\" | sed \"s|WORK_DIR|${work_dir}|g\" |sed \"s|JOB|${job}|g\"  > ${output_script}\n",
    "    sbatch ${output_script}\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fb50f31-c936-4bed-8df6-b88debf823f2",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "work_dir=\"/home/hs3393/cb_Mar/simulation_data/simulation_correlated/\"\n",
    "job=\"real_simulation_5trait_corr\"\n",
    "mkdir -p ${work_dir}\n",
    "mkdir -p ${work_dir}/code\n",
    "mkdir -p ${work_dir}/log\n",
    "cd ${work_dir}/code\n",
    "\n",
    "# Create the base_script file and write the bash code into it\n",
    "cat << 'EOF' > base_script\n",
    "#!/bin/bash -l\n",
    "# NOTE the -l flag!\n",
    "#\n",
    "#SBATCH -t 30:00:00\n",
    "#SBATCH --mem=30000\n",
    "#SBATCH -J JOB\n",
    "#SBATCH -o WORK_DIR/log/JOB.\"%j\".out\n",
    "#SBATCH -e WORK_DIR/log/JOB.\"%j\".err\n",
    "\n",
    "source ~/mamba_activate.sh\n",
    "module load Singularity\n",
    "\n",
    "\n",
    "sos run /home/hs3393/cb_Mar/simulation_code/7_Correlated_Simulation.ipynb JOB \\\n",
    "    --genofile `ls /home/hs3393/cloud_colocalization/simulation_data/selected_genes_genotype/*.bim` \\\n",
    "    --n_causal CAUSAL --mem 30G --h2g 0.05 --independent \\\n",
    "    --cwd WORK_DIR/\n",
    "EOF\n",
    "\n",
    "base_script=\"base_script\"\n",
    "for causal in 1 2 3 4; do\n",
    "    output_script=\"causal_${causal}.sh\"\n",
    "    cat ${base_script}|sed \"s|CAUSAL|${causal}|g\" | sed \"s|WORK_DIR|${work_dir}|g\" |sed \"s|JOB|${job}|g\"  > ${output_script}\n",
    "    sbatch ${output_script}\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c36e186-0c96-4b80-9052-9960217a6ab9",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "work_dir=\"/home/hs3393/cb_Mar/simulation_data/simulation_correlated/\"\n",
    "job=\"real_simulation_10trait_corr\"\n",
    "mkdir -p ${work_dir}\n",
    "mkdir -p ${work_dir}/code\n",
    "mkdir -p ${work_dir}/log\n",
    "cd ${work_dir}/code\n",
    "\n",
    "# Create the base_script file and write the bash code into it\n",
    "cat << 'EOF' > base_script\n",
    "#!/bin/bash -l\n",
    "# NOTE the -l flag!\n",
    "#\n",
    "#SBATCH -t 30:00:00\n",
    "#SBATCH --mem=30000\n",
    "#SBATCH -J JOB\n",
    "#SBATCH -o WORK_DIR/log/JOB.\"%j\".out\n",
    "#SBATCH -e WORK_DIR/log/JOB.\"%j\".err\n",
    "\n",
    "source ~/mamba_activate.sh\n",
    "module load Singularity\n",
    "\n",
    "\n",
    "sos run ./7_Correlated_Simulation.ipynb JOB \\\n",
    "    --genofile `ls /home/hs3393/cloud_colocalization/simulation_data/selected_genes_genotype/*.bim` \\\n",
    "    --n_causal CAUSAL --mem 30G --h2g 0.05 --independent \\\n",
    "    --cwd WORK_DIR/\n",
    "EOF\n",
    "\n",
    "base_script=\"base_script\"\n",
    "for causal in 1 2 3 4 5; do\n",
    "    output_script=\"causal_${causal}.sh\"\n",
    "    cat ${base_script}|sed \"s|CAUSAL|${causal}|g\" | sed \"s|WORK_DIR|${work_dir}|g\" |sed \"s|JOB|${job}|g\"  > ${output_script}\n",
    "    sbatch ${output_script}\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "R",
     "ir",
     "R",
     "",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.24.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
